File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/repository/ChannelRepository.java
package net.prostars.messagesystem.repository;

import jakarta.persistence.LockModeType;
import java.util.Optional;
import net.prostars.messagesystem.dto.projection.ChannelProjection;
import net.prostars.messagesystem.dto.projection.ChannelTitleProjection;
import net.prostars.messagesystem.dto.projection.InviteCodeProjection;
import net.prostars.messagesystem.entity.ChannelEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Repository;

@Repository
public interface ChannelRepository extends JpaRepository<ChannelEntity, Long> {

  Optional<ChannelTitleProjection> findChannelTitleByChannelId(@NonNull Long channelId);

  Optional<InviteCodeProjection> findChannelInviteCodeByChannelId(@NonNull Long channelId);

  Optional<ChannelProjection> findChannelByInviteCode(@NonNull String inviteCode);

  @Lock(LockModeType.PESSIMISTIC_WRITE)
  Optional<ChannelEntity> findForUpdateByChannelId(@NonNull Long channelId);
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/repository/ChannelRepository.kt
package com.yscorp.withpush.messagesystem.repository

import com.yscorp.withpush.messagesystem.dto.projection.ChannelProjection
import com.yscorp.withpush.messagesystem.dto.projection.ChannelTitleProjection
import com.yscorp.withpush.messagesystem.dto.projection.InviteCodeProjection
import com.yscorp.withpush.messagesystem.entity.ChannelEntity
import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository

@org.springframework.stereotype.Repository
interface ChannelRepository : JpaRepository<ChannelEntity, Long> {
    fun findChannelTitleByChannelId(channelId: Long): ChannelTitleProjection?

    fun findChannelInviteCodeByChannelId(channelId: Long): InviteCodeProjection?

    fun findChannelByInviteCode(inviteCode: String): ChannelProjection?

    @org.springframework.data.jpa.repository.Lock(LockModeType.PESSIMISTIC_WRITE)
    fun findForUpdateByChannelId(channelId: Long): ChannelEntity?

}
